### PR TITLE
Virtual Machine "raw" volume export proposal

### PR DESCRIPTION
**Pull request by mhenriks**
Signed-off-by: Michael Henriksen <mhenriks@redhat.com>

Supersedes: #155

### CODE REVIEW

File: design-proposals/export-import/volume-export.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+```
+
+```
+virtctl export delete snap1-export
+```
+
+Once a `VirtualMachineExport` is created, `virtctl export` can download volume archives
+
+```
+virtctl export download vm1-export --volume=volume1 --output volume1.tar.gz
```
Comment (ShellyKa13) [inline design-proposals/export-import/volume-export.md:151]: What about download a bundle of all the volumes of the vm?

Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:151]: yeah, I'll add that, the trick will be to do it in a way that preserves sparseness somehow.  Should be possible

File: design-proposals/export-import/volume-export.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+``` yaml
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-vm1
+spec:
+    source:
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
```
Comment (dankenigsberg) [inline design-proposals/export-import/volume-export.md:65]: What about VirtualMachineInstances? Would this support exporting those?

File: design-proposals/export-import/volume-export.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        volumes:
+        - name: kubevirt-volume
+          - format: raw
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/raw?token=8fmf94kf
+          - format: archive
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/archive?token=8fmf94kf
+        - name: archive-volume
+          - format: dir
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/archive-volume/dir?token=8fmf94kf
+          - format: archive
```
Comment (dankenigsberg) [inline design-proposals/export-import/volume-export.md:108]: I don't understand what are the different formats; Nor do I understand how would `status` look like if the VM has multiple disks, or once snapshots gain RAM.

Can you add an example of a multi-disk VM and how its VirtualMachineExport looks like? I'd like to ensure that I can easily recreate the VM on another cluster with access to its VirtualMachineExport alone.

File: design-proposals/export-import/volume-export.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+apiVersion: export.kubevirt.io/v1alpha1
+kind: VirtualMachineExport
+metadata:
+    name: export-vm1
+spec:
+    source:
+        apiGroup: kubevirt.io
+        kind: VirtualMachine
+        name: vm1
+```
```
Comment (dankenigsberg) [inline design-proposals/export-import/volume-export.md:67]: Idea (which may be postponed for phase 3 of the implementation): add an optional `target` section, which the vm owner can use to specify an S3 bucket and a secret to access it. This would allow me to export (an archive? of) the VM to the cloud, without having to go through my laptop.

File: design-proposals/export-import/volume-export.md [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+## Goals
+* Define VirtualMachineExport API
+* Introduce virtctl export command
+
+## Non Goals
+* Implementation details of internal components (servers/proxies/controllers/etc)
+
+## User Stories
+* As a KubeVirt user, I would like to export a Virtual Machine volume for local debugging/analysis.
```
Comment (dankenigsberg) [inline design-proposals/export-import/volume-export.md:15]: > export a Virtual Machine volume

Is this proposal about exporting a single volume? I think that the more interesting case is about exporting all VM data: its spec, multiple volumes and possibly RAM.

**Review by dankenigsberg — CHANGES_REQUESTED** [review]
thanks for the much-needed proposal.

File: design-proposals/export-import/volume-export.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+The http endpoint supports random access to a raw disk image.
+
+#### tar.gz
+
+The http endpoint is a link to a `tar.gz` streaming download.
+
+#### dir
+
+The http endpoint is a golang `http.FileServer` serving up the root of the filesystem.
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:53]: Basically serving up a `Filesystem` PVC from the host perspective.

File: design-proposals/export-import/volume-export.md [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+2.  Wait for the `Ready` condition of the VirtualMachineExport to be `True`
+3.  Download a Virtual Machine volume in the desired format listed under `status.links.volumes`
+
+### Export Formats
+
+| Volume Content Type | Export Format |
+| ------------------- | ------------- |
+| kubevirt            | raw           |
+| kubevirt            | tar.gz        |
+| archive             | dir           |
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:40]: DataVolumes support two different content types, `kubevirt` and `archive`.  The former results in a disk image at `/data/disk.img`.  The latter is simply the extracted contents of an archive so may be multiple files/directories.  Archive is IMO mostly useful for `virtiofs`.

File: design-proposals/export-import/volume-export.md [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+#### tar.gz
+
+The http endpoint is a link to a `tar.gz` streaming download.
+
+#### dir
+
+The http endpoint is a golang `http.FileServer` serving up the root of the filesystem.
+
+### Export a VirtualMachine
```
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:55]: If exporting a `VirtualMachineSnapshot`, first a PVC will be created from each `VolumeSnapshot` it contains.  Then, a pod will be started to serve up the PVC data.  When the `VirtualMachineExport` is deleted, the pod/PVCs will also be deleted.

If exporting a `VirtualMachine` it is expected that the VirtualMachine is stopped or in some sort of maintenance mode.  It should not be possible to export a running VM.

Yamls are already available via the k8s apiserver.

File: design-proposals/export-import/volume-export.md (hunk 2 above)
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:65]: Maybe?  Would probably want to enforce they are paused.  Also would need a way to start the export server in virt-launcher pod.  Perhaps this is something that could be added later

File: design-proposals/export-import/volume-export.md (hunk 3 above)
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:108]: THe example has multiple disks.  One named `kubevirt-volume` and one named `archive-volume`.

Snapshots with ram has not been designed yet, but I think the idea is that there will be a special "memory" volume for VMs that is snapshotted/restored just like the others.  I expect "memory" volumes will be exported in the "raw" format described here.  Maybe "file" is a better name for that format?

File: design-proposals/export-import/volume-export.md (hunk 5 above)
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:15]: The idea is that `VirtualMachineExport` will make it possible to export all persistent data (including mem dumps) related to a VirtualMachines.  Manifests are already available via k8s apiserver.  Could be added here but I think it would be better to rely on standard k8s API/RBAC for that.

**Review by davidvossel — COMMENTED** [review]
The technical details here with how the export occurs makes sense to me. The primary aspect of this design that I question is name and implications of the name `VirtualMachineExport` 

I typically think of import and export coming in pairs. If you export something, you should be able to import it as well to another cluster. Since a `VirtualMachineExport` only works with volumes, then the counterpart to this at the moment is a DataVolume import/upload, and not necessarily a future `VirtualMachineImport`

I find that mismatch odd from a consistency perspective. People export volumes using a VMExport, but import volumes one by one using a DataVolume import/upload.

If in the end we're only scoping this export to volumes (PVCs) then why don't we make the abstraction a logical counterpart to DataVolumes? Like a `DataVolumeExport` where the source is a PVC and the destination is S3, an http Download link, or even uploading the contents to a container registery.

Basically the inverse of a DataVolume.

File: design-proposals/export-import/volume-export.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:53]: For KubeVirt, isn't there a single raw file representing the disk?

Would you explain this in the text of the proposal? The purpose of this is not clear to me.

File: design-proposals/export-import/volume-export.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:40]: Thanks. I was not familiar with that. I would appreciate if this doc links to their respective definitions on CDI.

File: design-proposals/export-import/volume-export.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:55]: > If exporting a VirtualMachine it is expected that the VirtualMachine is stopped or in some sort of maintenance mode. It should not be possible to export a running VM.
>
> Yamls are already available via the k8s apiserver.

I worry that this may be raceful: after the VM is exported and while the data is downloaded, the VM may be started, modifying the data midway. Or the VM definition may be changed. What do you think about taking an implicit snapshot to avoid that?

File: design-proposals/export-import/volume-export.md [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+status:
+    phase: Ready # Pending|Ready|Terminated
+    token: 8fmf94kf
+    links:
+      internal:
+        cert: <base64 encoded cert>
+        volumes:
+        - name: kubevirt-volume
+          - format: raw
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/raw?token=8fmf94kf
```
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:102]: Thanks. What is the benefit of having it in its own field, too?

File: design-proposals/export-import/volume-export.md [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+          - format: archive
+            url: https://virt-import.kubevirt/ns1/export-vm1/volumes/kubevirt-volume/archive?token=8fmf94kf
```
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:104]: To me, a more natural name would be `virt-export` as the service is exporting VMs. Or do I miss something basic?

File: design-proposals/export-import/volume-export.md (hunk 3 above)
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:108]: > THe example has multiple disks. One named kubevirt-volume and one named archive-volume.

I did not realize that. I would appreciate seeing a skeleton of the VM being exported in the proposal.

>  Maybe "file" is a better name for that format?

I'm not sure about the context of the question. "raw" tells me that the file is not compressed or otherwise treated. "file tells me less.

File: design-proposals/export-import/volume-export.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md]: This goes back to the different contenttypes discussed elsewhere here.  I've put a little more info about `http.FileServer`.  But this is only used for `archive` contenttype

File: design-proposals/export-import/volume-export.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:44]: link/description added

File: design-proposals/export-import/volume-export.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:63]: I propose that we start with volume data and can further extend this resource to include manifests and/or a single ova like file containing all vm data and config.

File: design-proposals/export-import/volume-export.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md]: When serving up with `dir`, the token will be needed for accessing all urls but only the root url is listed in the VirtualMachineExport status

File: design-proposals/export-import/volume-export.md (hunk 3 above)
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md]: yeah raw is better

**Comment by mhenriks** [comment]
> The technical details here with how the export occurs makes sense to me. The primary aspect of this design that I question is name and implications of the name `VirtualMachineExport`
> 
> I typically think of import and export coming in pairs. If you export something, you should be able to import it as well to another cluster. Since a `VirtualMachineExport` only works with volumes, then the counterpart to this at the moment is a DataVolume import/upload, and not necessarily a future `VirtualMachineImport`
> 
> I find that mismatch odd from a consistency perspective. People export volumes using a VMExport, but import volumes one by one using a DataVolume import/upload.
> 
> If in the end we're only scoping this export to volumes (PVCs) then why don't we make the abstraction a logical counterpart to DataVolumes? Like a `DataVolumeExport` where the source is a PVC and the destination is S3, an http Download link, or even uploading the contents to a container registery.
> 
> Basically the inverse of a DataVolume.

cc @dankenigsberg @davidvossel 

I agree with much said here but there are longer term goals for `VirtualMachineExport` like including a link to download a single ova like file.  So I think it makes sense to work at the `VirtualMachine` level.

I have proposed adding a `PersistentVolumeClaim` source to `VirtualMachineExport` though.  That will make it easy to export a single PVC that may or may not be associated to a VM.

File: design-proposals/export-import/volume-export.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+
+```
+virtctl vmexport delete snap1-export
+```
+
+Once a `VirtualMachineExport` is created, `virtctl vmexport` can download volume archives
+
+```
+virtctl vmexport download vm1-export --volume=volume1 --output disk.img.gz
+```
```
Comment (davidvossel) [inline design-proposals/export-import/volume-export.md:174]: the virtctl image-upload command has the ability to both create a new datavolume and upload to it in a single command.

Could we do something similar for export, where someone can download a volume for a VM using a single command?

**Review by davidvossel — COMMENTED** [review]
/lgtm 

So, phase 1 is volume export, and a follow up phase to is full VM export (with spec and more). I'm alright with this as long as we keep the VirtualMachineExport api alpha until we've sorted out how the full VM export/import will work in the future.

File: design-proposals/export-import/volume-export.md (hunk 11 above)
Comment (mhenriks) [reply design-proposals/export-import/volume-export.md:174]: Yeah, would have to somehow configure the source/target cluster kubeconfigs somehow though

Comment (davidvossel) [reply design-proposals/export-import/volume-export.md:174]: i just meant the download action. Right now someone has to issue two commands `create` the export and `download` the volume

I was saying, can we streamline this and imply the create during the download, similar to how we imply the pvc create during the upload

File: design-proposals/export-import/volume-export.md (hunk 5 above)
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:15]: I'm not sure I understand your answer. The propose text speak about a (single) Virtual Machine volume. While I think it should speak about a virtual machine definition and all associated data.

File: design-proposals/export-import/volume-export.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (dankenigsberg) [reply design-proposals/export-import/volume-export.md:63]: It is not obvious to me how this would work. Would you lay out your phased approach in this proposal?

**Comment by rmohr** [comment]
/lgtm

similar thoughts to:

> /lgtm
> 
> So, phase 1 is volume export, and a follow up phase to is full VM export (with spec and more). I'm alright with this as long as we keep the VirtualMachineExport api alpha until we've sorted out how the full VM export/import will work in the future.

I think it is a start to experiment.
